### ISSUE TITLE
Exception in thread "main" java.lang.OutOfMemoryError: Java heap space

### ISSUE THREAD

**Issue by Intelligent2013** [metanorma/mn2pdf#214]
Source: https://github.com/metanorma/pdfa-iso-32000-2/pull/2

```
[INFO] Generation of Intermediate Format with information about the table's widths ...
[INFO] Rendering into intermediate format ...
This is a TrueType collection file with 2 fonts
Containing the following fonts:
Cambria <-- selected
Cambria Math
[INFO] Intermediate format. Rendered page #1.
[INFO] Intermediate format. Rendered page #2.
This is a TrueType collection file with 2 fonts
Containing the following fonts:
Cambria
Cambria Math <-- selected
This is a TrueType collection file with 2 fonts
Containing the following fonts:
Cambria
Cambria Math <-- selected
Exception in thread "main" java.lang.OutOfMemoryError: Java heap space
        at org.apache.fop.fonts.FontTriplet.getKey(FontTriplet.java:101)
        at org.apache.fop.fonts.FontTriplet.toString(FontTriplet.java:135)
        at org.apache.fop.fonts.FontTriplet.hashCode(FontTriplet.java:113)
        at java.util.HashMap.hash(HashMap.java:340)
        at java.util.HashMap.get(HashMap.java:558)
        at org.apache.fop.fonts.FontInfo.getInternalFontKey(FontInfo.java:527)
        at org.apache.fop.fonts.FontInfo.fontLookup(FontInfo.java:208)
        at org.apache.fop.fonts.FontInfo.fontLookup(FontInfo.java:386)
        at org.apache.fop.fonts.FontInfo.fontLookup(FontInfo.java:413)
        at org.apache.fop.fo.properties.CommonFont.getFontState(CommonFont.java:223)
        at org.apache.fop.fonts.FontSelector.selectFontForCharactersInText(FontSelector.java:110)
        at org.apache.fop.layoutmgr.inline.TextLayoutManager.processWord(TextLayoutManager.java:962)
        at org.apache.fop.layoutmgr.inline.TextLayoutManager.getNextKnuthElements(TextLayoutManager.java:822)
        at org.apache.fop.layoutmgr.inline.LineLayoutManager.collectInlineKnuthElements(LineLayoutManager.java:712)
        at org.apache.fop.layoutmgr.inline.LineLayoutManager.getNextKnuthElements(LineLayoutManager.java:630)
        at org.apache.fop.layoutmgr.BlockLayoutManager.getNextChildElements(BlockLayoutManager.java:141)
        at org.apache.fop.layoutmgr.BlockStackingLayoutManager.getNextKnuthElements(BlockStackingLayoutManager.java:292)
        at org.apache.fop.layoutmgr.BlockLayoutManager.getNextKnuthElements(BlockLayoutManager.java:113)
        at org.apache.fop.layoutmgr.BlockLayoutManager.getNextKnuthElements(BlockLayoutManager.java:105)
        at org.apache.fop.layoutmgr.table.TableCellLayoutManager.getNextKnuthElements(TableCellLayoutManager.java:191)
        at org.apache.fop.layoutmgr.table.RowGroupLayoutManager.createElementsForRowGroup(RowGroupLayoutManager.java:120)
        at org.apache.fop.layoutmgr.table.RowGroupLayoutManager.getNextKnuthElements(RowGroupLayoutManager.java:63)
        at org.apache.fop.layoutmgr.table.TableContentLayoutManager.getKnuthElementsForRowIterator(TableContentLayoutManager.java:284)
        at org.apache.fop.layoutmgr.table.TableContentLayoutManager.getNextKnuthElements(TableContentLayoutManager.java:212)
        at org.apache.fop.layoutmgr.table.TableLayoutManager.getNextKnuthElements(TableLayoutManager.java:277)
        at org.apache.fop.layoutmgr.BlockStackingLayoutManager.getNextChildElements(BlockStackingLayoutManager.java:501)
        at org.apache.fop.layoutmgr.BlockContainerLayoutManager.getNextKnuthElements(BlockContainerLayoutManager.java:256)
        at org.apache.fop.layoutmgr.BlockContainerLayoutManager.getNextKnuthElements(BlockContainerLayoutManager.java:186)
        at org.apache.fop.layoutmgr.BlockLayoutManager.getNextChildElements(BlockLayoutManager.java:141)
        at org.apache.fop.layoutmgr.BlockStackingLayoutManager.getNextKnuthElements(BlockStackingLayoutManager.java:292)
        at org.apache.fop.layoutmgr.BlockLayoutManager.getNextKnuthElements(BlockLayoutManager.java:113)
        at org.apache.fop.layoutmgr.BlockLayoutManager.getNextKnuthElements(BlockLayoutManager.java:105)
```

**Comment by Intelligent2013** [comment]
The error `java.lang.OutOfMemoryError: Java heap space` occurs due the intermediate XSL-FO (for table auto-layout algorithm) big file size (49Mb).
The quick solution is increasing Java heap size up to 3Gb (currently 2Gb) : `-Xmx3072m`
In `mn2pdf-ruby`
https://github.com/metanorma/mn2pdf-ruby/blob/e09dff5ce62031536c32184f9af96175fe56d25b/lib/mn2pdf.rb#L8

After that the PDF creates successfully.

Another solution: optimize intermediate XSL-FO generation (and/or auto-layout algorithm). I'll investigate how to do this.

**Comment by Intelligent2013** [comment]
The error occurs in the StringBuilder to String conversion:
```java
    public String getResultedXML() {
        return sbResult.toString();
    }
```
The size of _sbResult_ is 57,275,878 chars.

**Comment by Intelligent2013** [comment]
Somewhere there is a memory leaks:
![image](https://github.com/metanorma/mn2pdf/assets/5559926/818e5471-24ed-463c-b065-fab22c460f90)

Java Heap size is 700Mb before starting of the transforming into the IF XML on the post-processing step.

**Comment by Intelligent2013** [comment]
Memory cleanup added (non using resources removed). Java heap size in 365Mb now:
![image](https://github.com/metanorma/mn2pdf/assets/5559926/a69410fa-dcbf-4ebd-bbb6-4315dd4ff7b1)

But again error occurs:
```
Exception in thread "main" java.lang.OutOfMemoryError: Java heap space
	at java.util.Arrays.copyOfRange(Arrays.java:3664)
	at java.lang.String.<init>(String.java:207)
	at java.lang.StringBuilder.toString(StringBuilder.java:413)
	at org.metanorma.fop.ifhandler.FOPIFHiddenMathHandler.getResultedXML(FOPIFHiddenMathHandler.java:262)
```


Java used memory size: 1569347912
Java available memory size: 339584696
sbResult (StringBuilder) length: 57275878

Very strange that StringBuilder (57Mb) can't be converted into String with  339Mb available memory.

**Comment by Intelligent2013** [comment]
StringBuilder toString() moved to the external class - it solved the issue.

**Comment by Intelligent2013** [comment]
PDF didn't generated by:
```
bundle exec metanorma -t iso -x pdf document.adoc 
```

![image](https://github.com/metanorma/mn2pdf/assets/5559926/e18c332b-6c71-4e7d-8f23-1fb657b4222b)


The command line is:
```
java -Xss5m -Xmx2048m -Djava.awt.headless=true -Duser.home=C:/Users/TestUser -jar C:/tools/ruby31/lib/ruby/gems/3.1.0/gems/mn2pdf-1.74/lib/../bin/mn2pdf.jar --xml-file "C:/Users/TestUser/AppData/Local/Temp/document20230626-15288-pudz7s.xml" --xsl-file "C:/tools/ruby31/lib/ruby/gems/3.1.0/gems/metanorma-iso-2.4.6/lib/isodoc/iso/iso.international-standard.xsl" --pdf-file "D:/Work/Metanorma/repositories/pdfa-iso-32000-2/sources/iso-32000-2-202x/document.pdf" --syntax-highlight  --font-manifest "C:/Users/TestUser/AppData/Local/Temp/fontist_locations20230626-15288-w8t9bi.yml"
```

**Comment by Intelligent2013** [comment]
The command
`bundle exec metanorma -t iso -x pdf document.adoc`
uses ISO XSLT from latest gem release, but optimized XSLT updated later. 

Without ISO XSLT optimization:
![image](https://github.com/metanorma/mn2pdf/assets/5559926/3985508a-381f-40cc-bc68-daecc9ea4396)

With ISO XSLT optimization:
![image](https://github.com/metanorma/mn2pdf/assets/5559926/3c8288b1-037c-4d69-bad9-b1e831c8d9ea)


So just need update `Gemfile`:
```
gem "metanorma-iso", git: "https://github.com/metanorma/metanorma-iso", branch: "main"
```
and then execute `bundle update`.
In any case, I've made one more optimization in `mn2pdf` - preliminary removing 'semantic` part from Presentation XML. The source XML size decreased from 42Mb to 20Mb, and therefore memory consumption also optimized.
https://github.com/metanorma/mn2pdf/releases/tag/v1.75

**Comment by Intelligent2013** [comment]
Today the issue occurs again:
```
Java used memory size after: 1580713808
Java available memory size after: 328218800
Exception in thread "main" java.lang.OutOfMemoryError: Java heap space
        at java.util.Arrays.copyOfRange(Arrays.java:3664)
        at java.lang.String.<init>(String.java:207)
        at java.lang.StringBuilder.toString(StringBuilder.java:413)
        at org.metanorma.fop.PDFGenerator.addHiddenMath(PDFGenerator.java:860)
        at org.metanorma.fop.PDFGenerator.runFOP(PDFGenerator.java:530)
        at org.metanorma.fop.PDFGenerator.convertmn2pdf(PDFGenerator.java:483)
        at org.metanorma.fop.PDFGenerator.process(PDFGenerator.java:310)
        at org.metanorma.fop.mn2pdf.main(mn2pdf.java:350)
```
After small modification of the XSLT for https://github.com/metanorma/mn2pdf/issues/218.
It occurs in 50% cases on the same input data.

**Comment by Intelligent2013** [comment]
Memory consumption optimized.
